### PR TITLE
Deploy solver

### PR DESCRIPTION
**Pull request by shlomitk1**
Fixes https://github.com/fybrik/fybrik/issues/1260
A new section was added to manager values yaml:
  ```
# CSP solver for data plane optimization
  solver:
    # image of the container with solver binary and libs
    # when specified, the solver will be deployed in the manager pod
    image: "optimizer"
    imagePullPolicy: "Always"
    # Set to true to enable the use of the solver by Fybrik
    enabled: false

```
If an image including solver binary and libs is specified, an init container runs that copies solver files from the image to the manager container.

### CODE REVIEW

File: charts/fybrik/values.yaml [hunk 1]
```diff
@@ -191,6 +191,15 @@ manager:
...
   prometheus: false
 
+  # CSP solver for data plane optimization
+  solver:
+    # image of the container with solver binary and libs
+    # when specified, the solver will be deployed in the manager pod
+    image: "optimizer"
+    imagePullPolicy: "Always"
+    # Set to true to enable the use of the solver by Fybrik
+    enabled: false
```
Comment (simanadler) [inline charts/fybrik/values.yaml:201]: The default is to deploy the solver but to have it turned off?  Shouldn't the default be to have it turned on?

File: pkg/optimizer/optimizer.go [hunk 2]
```diff
@@ -45,18 +46,20 @@ func NewOptimizer(env *datapath.Environment, problemData *datapath.DataInfo, sol
 }
 
 func (opt *Optimizer) getSolution(pathLength int) (string, error) {
+	opt.log.Debug().Msgf("finding solution of length %d", pathLength)
```
Comment (simanadler) [inline pkg/optimizer/optimizer.go:49]: Why not use the standardized fybrik logging?  Same comment for other logging calls in this file.

**Review by simanadler — COMMENTED** [review]
See comments

File: pkg/optimizer/optimizer.go (hunk 2 above)
Comment (shlomitk1) [reply pkg/optimizer/optimizer.go:49]: It is standardized. "log" is a member of Optimizer which comes from "github.com/rs/zerolog"
This printout will look like
`{"level":"debug","Controller":"FybrikApplication","FybrikApplication":"default/test-app-prod","app.fybrik.io/app-uuid":"b848eb60-8d2a-4d58-89b3-9d3e8197978b","time":"2022-05-30T06:57:27Z","caller":"/data/ws/the-mesh-for-data/pkg/optimizer/optimizer.go:49","message":"finding solution of length 1"}`

File: charts/fybrik/values.yaml (hunk 1 above)
Comment (shlomitk1) [reply charts/fybrik/values.yaml:201]: I still consider this as an experimental release. If it goes into 0.7 it should definitely be "false". Later we'll change it to "true", after more runs and features.

Comment (simanadler) [reply charts/fybrik/values.yaml:201]: OK.  Makes sense

File: pkg/optimizer/Makefile [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+
+# Overwrite docker-build from docker.mk
+.PHONY: docker-build
+docker-build: 	
+	mkdir solver-tools
+	mkdir solver-tools/bin
+	mkdir solver-tools/lib
+	cp ${ABSTOOLBIN}/fzn-or-tools solver-tools/bin/solver
+	cp ${ABSTOOLBIN}/../lib/libfz.so solver-tools/lib/
+	cp ${ABSTOOLBIN}/../lib/libortools.so solver-tools/lib/
```
Comment (zivnevo) [inline pkg/optimizer/Makefile:20]:
```suggestion
	cp ${ABSTOOLBIN}/fzn-or-tools solver-tools/bin/solver
	cp -r ${ABSTOOLBIN}/../lib solver-tools
```
It is not guaranteed that in the coming versions of `fzn-or-tools` we'll have exactly these 2 libraries (in fact, the next release seems to have a different name for these libraries)

File: pkg/optimizer/Makefile [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+KUBE_NAMESPACE ?= fybrik-system
+CONTROLLER_NAMESPACE ?= ${KUBE_NAMESPACE}
+
+include $(ROOT_DIR)/hack/make-rules/tools.mk
+include $(ROOT_DIR)/hack/make-rules/docker.mk
+include $(ROOT_DIR)/hack/make-rules/verify.mk
+
+# Overwrite docker-build from docker.mk
+.PHONY: docker-build
+docker-build:
```
Comment (zivnevo) [inline pkg/optimizer/Makefile:14]: Don't we also need a rule for `docker-push`?

File: charts/fybrik/templates/manager-deployment.yaml [hunk 5]
```diff
@@ -34,6 +34,20 @@ spec:
       serviceAccountName: {{ .Values.manager.serviceAccount.name }}
       securityContext:
         {{- toYaml .Values.manager.podSecurityContext | nindent 8 }}
+      {{- if .Values.manager.solver.image }}
```
Comment (zivnevo) [inline charts/fybrik/templates/manager-deployment.yaml:37]:
```suggestion
      {{- if .Values.manager.solver.enabled}}
```
It's the `enabled` flag we want to depend on, isn't it?
Note that this `if` statement appears 3 more times

**Review by zivnevo — CHANGES_REQUESTED** [review]
Very neat and :cool:
See some small comments below

File: charts/fybrik/templates/manager-deployment.yaml (hunk 5 above)
Comment (shlomitk1) [reply charts/fybrik/templates/manager-deployment.yaml:37]: We want to deploy once and be able to switch usage on the fly by changing the configmap manually.

File: pkg/optimizer/Makefile (hunk 3 above)
Comment (shlomitk1) [reply pkg/optimizer/Makefile:20]: ok, will change, thanks.

**Review by zivnevo — APPROVED** [review]

File: pkg/optimizer/optimizer.go (hunk 2 above)
Comment (simanadler) [reply pkg/optimizer/optimizer.go:49]: OK

**Comment by roytman** [comment]
converted to draft to prevent merging it into v0.7.

**Pull request merged**